File: web/src/app/services/demande.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, from, throwError } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import { Auth } from '@angular/fire/auth';
import { Demande } from '../models/demande.model'; // Make sure to have a 'Demande' model

@Injectable({
  providedIn: 'root',
})
export class DemandeService {
  private apiUrl = 'http://localhost:8888/demande-service/api/demandes'; // Use your actual URL

  constructor(private http: HttpClient, private auth: Auth) {}

  private getHeaders(): Observable<HttpHeaders> {
    return this.auth.currentUser
      ? from(this.auth.currentUser.getIdToken()).pipe(
          map((token) =>
            new HttpHeaders().set('Authorization', `Bearer ${token}`)
          )
        )
      : throwError(() => new Error('User is not logged in'));
  }

  getAllDemandes(): Observable<Demande[]> {
    return this.getHeaders().pipe(
      switchMap((headers) => this.http.get<Demande[]>(this.apiUrl, { headers }))
    );
  }

  createDemande(demande: Demande): Observable<Demande> {
    return this.getHeaders().pipe(
      switchMap((headers) =>
        this.http.post<Demande>(this.apiUrl, demande, { headers })
      )
    );
  }

  updateDemande(demande: Demande): Observable<Demande> {
    return this.getHeaders().pipe(
      switchMap((headers) =>
        this.http.put<Demande>(`${this.apiUrl}/${demande.id}`, demande, {
          headers,
        })
      )
    );
  }

  deleteDemande(demandeId: number): Observable<any> {
    return this.getHeaders().pipe(
      switchMap((headers) =>
        this.http.delete(`${this.apiUrl}/${demandeId}`, { headers })
      )
    );
  }
}

File: mobile/lib/services/demande_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:firebase_auth/firebase_auth.dart';

class DemandeService {
  final String apiUrl =
      'https://e148-196-89-158-31.ngrok-free.app/demande-service/api/demandes';
  final FirebaseAuth _auth = FirebaseAuth.instance;

  Future<String> _getIdToken() async {
    return await _auth.currentUser!.getIdToken() ?? "";
  }

  Future<List<dynamic>> getAllDemandes() async {
    String token = await _getIdToken();
    final response = await http.get(
      Uri.parse(apiUrl),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load demandes');
    }
  }

  Future<dynamic> createDemande(Map<String, dynamic> demande) async {
    String token = await _getIdToken();
    final response = await http.post(
      Uri.parse(apiUrl),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token'
      },
      body: json.encode(demande),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to create demande');
    }
  }

  Future<dynamic> updateDemande(int id, Map<String, dynamic> demande) async {
    String token = await _getIdToken();
    final response = await http.put(
      Uri.parse('$apiUrl/$id'),
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $token'
      },
      body: json.encode(demande),
    );

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to update demande');
    }
  }

  Future<void> deleteDemande(int id) async {
    String token = await _getIdToken();
    final response = await http.delete(
      Uri.parse('$apiUrl/$id'),
      headers: {'Authorization': 'Bearer $token'},
    );

    if (response.statusCode != 200) {
      throw Exception('Failed to delete demande');
    }
  }
}
